agents/ppo.py

```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        )
+
         if self._bag_trajectory_memory is not None:
             # TODO[Vladimir Baikalov]: Think about how to generalize
             # Shape: [batch_size] if exists, None otherwise
-            bag_ids = inputs.get(self._bag_idx_prefix, None)
+            bag_ids = inputs.get(self._bag_ids_prefix, None)
 
             self._bag_trajectory_memory.add_to_trajectory(
                 bag_ids=bag_ids,
-                node_idxs=current_node_idx,
+                node_idxs=torch.full([batch_size], self._node_id),
                 extra_infos=zip(
-                    current_state_value_function.detach(),
-                    neighbors_logits.detach(),
-                    bag_ids,
-                    current_node_idx,
+                    torch.unsqueeze(current_state_value_function.detach(), dim=1),
+                    torch.unsqueeze(current_node_idx, dim=1),
                     neighbor_node_ids,
-                    destination_node_idx
+                    torch.unsqueeze(next_neighbor_ids.detach(), dim=1),
+                    torch.unsqueeze(neighbors_logits.detach(), dim=1),
+                    torch.unsqueeze(destination_node_idx, dim=1),
+                    torch.unsqueeze(next_state_value_function.detach(), dim=1),
                 )
             )
 
-        return {
-            'predicted_next_node_idx': next_neighbors_ids,
+        inputs[self._output_prefix] = next_neighbor_ids
+        inputs.update({
+            'predicted_next_node_idx': next_neighbor_ids,
             'predicted_next_node_logits': neighbors_logits,
             'predicted_current_state_v_value': current_state_value_function
-        }
-
-    def learn(self):
-        for trajectory in self._bag_trajectory_memory.sample_trajectories_for_node_idx(
-                self._node_idx,
-                self._trajectory_sample_size
-        ):
-            loss = self._trajectory_loss(trajectory)
-            # TODO backward here? or return to env?
-
-    def _trajectory_loss(self, trajectory):
-        reward = [data['reward'] for data in trajectory]
-        start_v_func, policy, node_idx, neighbor, destination, storage = trajectory[0]['extra_infos']
-        end_v_func = trajectory[-1]['extra_infos'][0]
-        return self._loss(node_idx, neighbor, destination, policy, start_v_func, reward, end_v_func)
+        })
+        # print(current_node_idx, destination_node_idx, current_state_value_function.detach())
+        # print(neighbor_node_ids.padded_values, neighbors_logits.detach())
+        return inputs
+
+    def learn(self) -> Optional[Tensor]:
+        loss = 0
+        learn_trajectories = self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+            self._node_id,
+            self._trajectory_sample_size,
+            self._trajectory_length
+        )
+        if not learn_trajectories:
+            return None
+        learn_norm_rewards = get_norm_rewards(learn_trajectories)
```

Comment:
It is better to do special condition for this

Reply:
maybe make it todo? I don't sure yet what kind of loss transform make sense, it is in pre-alpha =)

Reply:
Yeah, let's make here TODO for now